### DISCUSSION TITLE
How would I send a message from another thread?

### DISCUSSION THREAD

**Discussion by Daanish-Khan** [Rapptz/discord.py#5868] [Q&A]
I've got a bot that sends live updates to a channel. I've created a listener to obtain data, which is then processed and sent into it's respective discord channel. However, I am unable to send messages as it needs async, and creating a thread with async isn't working.

Asyncio has also proven to be difficult. I've tried asyncio.run_coroutine_threadsafe, which for some reason prevents the bot from processing commands. Creating a new event loop gives me `Timeout context manager should be used inside a task.`

Any ideas?

This is code below that is similar to what I have at the moment.

```python

async def process_message(ctx):
    while True:
        if condition:
            continue
        else:
            # processing
            for key in channeldict:
                if key in title:
                    await ctx.guild.get_channel(channeldict[key]).send("test")

@client.command()
async def start_listening(ctx):
    thread = threading.Thread(target=process_message, args=[ctx])
    thread.start()

**Comment by Rapptz** [comment]
Please see the [background task example](https://github.com/Rapptz/discord.py/blob/master/examples/background_task.py) for how to do this, alternatively if you need this to be wrapped in a prettier interface there's the [`ext.tasks`](https://discordpy.readthedocs.io/en/latest/ext/tasks/index.html) extension that comes with the library.

**Reply by Daanish-Khan** [reply]
I've tried something similar with `create_task`, however it prevents the other commands from executing for whatever reason. I'll try using the example but I'm doubtful it works.

**Reply by Rapptz** [reply]
The reason why it's halting is because if `condition` is `True` in your example then it leads to an infinite loop without yielding back to the caller, see [What does "blocking" mean?](https://discordpy.readthedocs.io/en/latest/faq.html#what-does-blocking-mean) for more information there.

In order to play nicely with the asyncio system you have to yield back to the caller by using an `await` expression. So essentially:

```py
import asyncio

async def never_ending():
    condition = False
    while True:
        if condition:
            # Never ending loop
            continue

        await do_work()
        condition = True
```

This function will never yield back to the caller when `condition = True` because it's a never ending loop.

If you want to do some co-operative work queue with a task you'll have to be a bit more involved using an [`asyncio.Queue`](https://docs.python.org/3/library/asyncio-queue.html#asyncio.Queue). Note that this is true even if you use threading. Unfortunately I do not know enough about your situation to give you further assistance. However that's the crux of the issue here.
